File: Magic Squares In Grid/Magic Squares In Grid.py
# Solution for Magic Squares In Grid in PY

class Solution:
    def isMagicSquare(self, grid: List[List[int]], start_row: int, start_col: int) -> bool:
        # Ensure all numbers are distinct and in range [1, 9]
        unique_values = set()

        for r in range(3):
            for c in range(3):
                value = grid[start_row + r][start_col + c]
                if value < 1 or value > 9 or value in unique_values:
                    return False
                unique_values.add(value)

        # Target sum from first row
        target_sum = (
            grid[start_row][start_col] +
            grid[start_row][start_col + 1] +
            grid[start_row][start_col + 2]
        )

        # Check rows and columns
        for i in range(3):
            row_sum = (
                grid[start_row + i][start_col] +
                grid[start_row + i][start_col + 1] +
                grid[start_row + i][start_col + 2]
            )

            col_sum = (
                grid[start_row][start_col + i] +
                grid[start_row + 1][start_col + i] +
                grid[start_row + 2][start_col + i]
            )

            if row_sum != target_sum or col_sum != target_sum:
                return False

        # Check diagonals
        main_diagonal = (
            grid[start_row][start_col] +
            grid[start_row + 1][start_col + 1] +
            grid[start_row + 2][start_col + 2]
        )

        anti_diagonal = (
            grid[start_row][start_col + 2] +
            grid[start_row + 1][start_col + 1] +
            grid[start_row + 2][start_col]
        )

        return main_diagonal == target_sum and anti_diagonal == target_sum

    def numMagicSquaresInside(self, grid: List[List[int]]) -> int:
        total_rows = len(grid)
        total_cols = len(grid[0])
        magic_square_count = 0

        # Slide a 3x3 window over the grid
        for row in range(total_rows - 2):
            for col in range(total_cols - 2):
                if self.isMagicSquare(grid, row, col):
                    magic_square_count += 1

        return magic_square_count

File: Magic Squares In Grid/Magic Squares In Grid.java
// Solution for Magic Squares In Grid in JAVA

class Solution {

    /**
     * Checks whether the 3x3 subgrid starting at (startRow, startCol)
     * forms a valid magic square.
     */
    private boolean isMagicSquare(int[][] grid, int startRow, int startCol) {

        // To ensure all numbers are distinct and in the range [1, 9]
        Set<Integer> uniqueValues = new HashSet<>();

        // Validate number range and uniqueness
        for (int rowOffset = 0; rowOffset < 3; rowOffset++) {
            for (int colOffset = 0; colOffset < 3; colOffset++) {

                int value = grid[startRow + rowOffset][startCol + colOffset];

                // Magic square must contain numbers from 1 to 9 exactly once
                if (value < 1 || value > 9 || uniqueValues.contains(value)) {
                    return false;
                }
                uniqueValues.add(value);
            }
        }

        // Target sum using the first row
        int targetSum =
            grid[startRow][startCol] +
            grid[startRow][startCol + 1] +
            grid[startRow][startCol + 2];

        // Check all rows and columns
        for (int i = 0; i < 3; i++) {

            int rowSum =
                grid[startRow + i][startCol] +
                grid[startRow + i][startCol + 1] +
                grid[startRow + i][startCol + 2];

            int colSum =
                grid[startRow][startCol + i] +
                grid[startRow + 1][startCol + i] +
                grid[startRow + 2][startCol + i];

            if (rowSum != targetSum || colSum != targetSum) {
                return false;
            }
        }

        // Check both diagonals
        int mainDiagonal =
            grid[startRow][startCol] +
            grid[startRow + 1][startCol + 1] +
            grid[startRow + 2][startCol + 2];

        int antiDiagonal =
            grid[startRow][startCol + 2] +
            grid[startRow + 1][startCol + 1] +
            grid[startRow + 2][startCol];

        return mainDiagonal == targetSum && antiDiagonal == targetSum;
    }

    /**
     * Counts the number of 3x3 magic square subgrids.
     */
    public int numMagicSquaresInside(int[][] grid) {

        int totalRows = grid.length;
        int totalCols = grid[0].length;
        int magicSquareCount = 0;

        // Slide a 3x3 window over the grid
        for (int row = 0; row <= totalRows - 3; row++) {
            for (int col = 0; col <= totalCols - 3; col++) {
                if (isMagicSquare(grid, row, col)) {
                    magicSquareCount++;
                }
            }
        }

        return magicSquareCount;
    }
}
